### ISSUE TITLE
不知能否可以告知SVTRv2论文中的热力图是如何生成的？

### ISSUE THREAD

**Issue by JingWang-cloud** [Topdu/OpenOCR#90]
论文第三页和第五页关于SGM的部分有绘制热力图，不知是否可以告知这是如何生成的？十分感谢！

**Comment by Topdu** [comment]
参考我们开源的SMTR可视化代码vis_attn_smtr.py和vis_attn_smtr_bi.py，下载链接：https://drive.google.com/drive/folders/1dCuaWwCLP9xIHgy-7NtpeDLOvgk9NoKE

**Comment by ninhngocnhi** [comment]
@Topdu  I have 2 problems when using self.attn_maps in class SMTRDecoder. First I want to infer batch to reduce time infer but the model returns 1 attn_maps shared with all images in batch, it is not correct. Second, I convert onnx when make class GCTDecoder returns ctc_pred, self.gtc_decoder.attn_maps, it can't complete with lots of bugs and errors relevant to self.attn_maps. Can you help me please. Some logs when convert onnx like that:   
  %436799 : Float(16, 2, 1, 192, strides=[384, 192, 384, 1], requires_grad=1, device=cpu) = aten::mul(%431897, %438157), scope: openrec.modeling.base_recognizer.BaseRecognizer::/openrec.modeling.decoders.GTCDecoder::decoder/openrec.modeling.decoders.smtr_decoder.SMTRDecoder::gtc_decoder/openrec.modeling.decoders.smtr_decoder.SSMatchLayer::cmff_decoder.0/openrec.modeling.decoders.smtr_decoder.CrossAttention::question_to_images_cross_attn # OpenOCR/openrec/modeling/decoders/smtr_decoder.py:41:0
  %attn : Float(16, 2, 1, 1260, strides=[2520, 1260, 1260, 1], requires_grad=1, device=cpu) = aten::matmul(%436799, %355850), scope: openrec.modeling.base_recognizer.BaseRecognizer::/openrec.modeling.decoders.GTCDecoder::decoder/openrec.modeling.decoders.smtr_decoder.SMTRDecoder::gtc_decoder/openrec.modeling.decoders.smtr_decoder.SSMatchLayer::cmff_decoder.0/openrec.modeling.decoders.smtr_decoder.CrossAttention::question_to_images_cross_attn # OpenOCR/openrec/modeling/decoders/smtr_decoder.py:48:0
  %input.4369 : Float(16, 2, 1, 1260, strides=[2520, 1260, 1260, 1], requires_grad=1, device=cpu) = aten::softmax(%attn, %437358, %353643), scope: openrec.modeling.base_recognizer.BaseRecognizer::/openrec.modeling.decoders.GTCDecoder::decoder/openrec.modeling.decoders.smtr_decoder.SMTRDecoder::gtc_decoder/openrec.modeling.decoders.smtr_decoder.SSMatchLayer::cmff_decoder.0/openrec.modeling.decoders.smtr_decoder.CrossAttention::question_to_images_cross_attn # venv/lib/python3.10/site-packages/torch/nn/functional.py:1888:0
  %431929 : Float(16, 2, 1, 1260, strides=[2520, 1260, 1260, 1], requires_grad=1, device=cpu) = aten::dropout(%input.4369, %437360, %437350), scope: openrec.modeling.base_recognizer.BaseRecognizer::/openrec.modeling.decoders.GTCDecoder::decoder/openrec.modeling.decoders.smtr_decoder.SMTRDecoder::gtc_decoder/openrec.modeling.decoders.smtr_decoder.SSMatchLayer::cmff_decoder.0/openrec.modeling.decoders.smtr_decoder.CrossAttention::question_to_images_cross_attn/torch.nn.modules.dropout.Dropout::attn_drop # venv/lib/python3.10/site-packages/torch/nn/functional.py:1295:0
  %431949 : Float(2, 1, 1260, strides=[1260, 1260, 1], requires_grad=1, device=cpu) = aten::select(%431929, %437362, %437362), scope: openrec.modeling.base_recognizer.BaseRecognizer::/openrec.modeling.decoders.GTCDecoder::decoder/openrec.modeling.decoders.smtr_decoder.SMTRDecoder::gtc_decoder # OpenOCR/openrec/modeling/decoders/smtr_decoder.py:598:0
  return (%355567, %355877, %356173, %356469, %356765, %357061, %357357, %357653, %357949, %358245, %358541, %358837, %359133, %359429, %359725, %360021, %360317, %360613, %360909, %361205, %361501, %361797, %362093, %362389, %362685, %362981, %363277, %363573, %363869, %364165, %364461, %364757, %365053, %365349, %365645, %365941, %366237, %366533, %366829, %367125, %367421, %367717, %368013, %368309, %368605, %368901, %369197, %369493, %369789, %370085, %370381, %370677, %370973, %371269, %371565, %371861, %372157, %372453, %372749, %373045, %373341, %373637, %373933, %374229, %374525, %374821, %375117, %375413, %375709, %376005, %376301, %376597, %376893, %377189, %377485, %377781, %378077, %378373, %378669, %378965, %379261, %379557, %379853, %380149, %380445, %380741, %381037, %381333, %381629, %381925, %382221, %382517, %382813, %383109, %383405, %383701, %383997, %384293, %384589, %384885, %385181, %385477, %385773, %386069, %386365, %386661, %386957, %387253, %387549, %387845, %388141, %388437, %388733, %389029, %389325, %389621, %389917, %390213, %390509, %390805, %391101, %391397, %391693, %391989, %392285, %392581, %392877, %393173, %393469, %393765, %394061, %394357, %394653, %394949, %395245, %395541, %395837, %396133, %396429, %396725, %397021, %397317, %397613, %397909, %398205, %398501, %398797, %399093, %399389, %399685, %399981, %400277, %400573, %400869, %401165, %401461, %401757, %402053, %402349, %402645, %402941, %403237, %403533, %403829, %404125, %404421, %404717, %405013, %405309, %405605, %405901, %406197, %406493, %406789, %407085, %407381, %407677, %407973, %408269, %408565, %408861, %409157, %409453, %409749, %410045, %410341, %410637, %410933, %411229, %411525, %411821, %412117, %412413, %412709, %413005, %413301, %413597, %413893, %414189, %414485, %414781, %415077, %415373, %415669, %415965, %416261, %416557, %416853, %417149, %417445, %417741, %418037, %418333, %418629, %418925, %419221, %419517, %419813, %420109, %420405, %420701, %420997, %421293, %421589, %421885, %422181, %422477, %422773, %423069, %423365, %423661, %423957, %424253, %424549, %424845, %425141, %425437, %425733, %426029, %426325, %426621, %426917, %427213, %427509, %427805, %428101, %428397, %428693, %428989, %429285, %429581, %429877, %430173, %430469, %430765, %431061, %431357, %431653, %431949)
, {'input': {0: 'batch_size', 2: 'in_width', 3: 'int_height'}, 'ctc_pred': {0: 'batch_size', 2: 'out_width', 3: 'out_height'}, 'attn': {'0': 'token', '1': 'attn_map'}}, ['input']
